### ISSUE TITLE
SFT recipe eval: /forward requires next-token labels, but the recipe renders unshifted ones

### ISSUE THREAD

**Issue by sfc-gh-thonguyen** [Snowflake-AI-Research/Arctic-Platform#98]
Splitting this out of the work to make the Cortex/Neutrino `forward` operation
functional (see Dependencies below). Moving an SFT eval from forward-backward to
`forward` is not a drop-in substitution: the two endpoints can disagree about who
shifts labels, and the failure mode is silent — no error, just wrong numbers.

**There is a clean way to avoid the whole problem: send
`processing={"loss_fn": "causal_cross_entropy"}` so both endpoints run the same
server-side code path.** See [Recommended fix](#recommended-fix).

## The asymmetry

`forward` gathers per-token log-probabilities straight out of `logits[t]` and
**never shifts**, so it requires next-token alignment
(`labels[t] == input_ids[t+1]`). On the dss-platform side:

- `compute_logprobs_post` gathers at `context["labels"]` with no shifting
  (`arctic_training/arctic_rl/processors/pipeline.py`)
- the worker rejects a batch without labels outright: *"forward_only requires
  next-token labels created before packing"* (`dss/ray_dss/jobs/gpu/training.py`)
- `ensure_next_token_labels` is a no-op when `labels` is already present, which it
  always is for SFT

Plain SFT forward-backward is different, and provider-dependent. In
`dp_shard_batch` (`dss/ray_dss/jobs/gpu/sp/data_plane.py`):

```python
is_sft = attach_global_loss_counts and not isinstance(batch.get("processing"), dict)
request_labels_are_shifted = sft_labels_are_shifted        # True iff model_provider == "prime_rl"
if is_sft:
    if sft_labels_are_shifted:            # prime_rl: labels already next-token
        ensure_next_token_labels(batch)   # no-op when labels are present
    elif "labels" not in batch:
        raise ValueError("labels-only SFT requests require tensor labels")
    else:                                  # HF / Liger: shift them here
        shift_sft_labels_for_logits(batch)
        request_labels_are_shifted = True
```

`shift_sft_labels_for_logits` (`dss/ray_dss/jobs/gpu/sp/loss.py`) does the shift
before packing, because a model-side internal shift would take the next packed
row's first token as this row's last target.

So on the **plain-SFT path** with an HF/Liger job, one batch needs two different
label tensors depending on which endpoint it goes to.

## Why the recipe hits it

In the conversational SFT recipe being added in #88
(`cortex_training/recipes/sft/conversational/train.py`), `eval_nll` calls
`forward_loss`, which is an alias for `forward_backward`, with no `processing` — so
it takes the plain-SFT path. Its labels come from:

```python
next_token_labels = use_next_token_labels(config.model_provider)
```

and `use_next_token_labels` returns `True` only for `prime_rl`. The default is
`model_provider = "huggingface"`, so `sequence_from_conversation` emits
`labels[t] = input_ids[t]` — correct for plain-SFT forward-backward (the server
shifts it), wrong for `forward`.

Feed those to `forward` and you gather `log P(input_ids[t] | prefix through t)` —
the probability of a token the prefix already contains. Nothing on either side
validates the alignment, so it returns plausible-looking numbers.

## Recommended fix

Rather than maintaining two label conventions, **route the supervised loss through
the RL pipeline** by passing a `processing` dict:

```python
processing = {"loss_fn": "causal_cross_entropy", "post": ["compute_logprobs"], "config": {}}
```

That makes `is_sft` False, so `dp_shard_batch` skips the provider-dependent
`shift_sft_labels_for_logits` branch entirely, and `fwd_bwd` dispatches to
`_rl_fwd_bwd` → `run_pipeline` — the same machinery `forward_only` already uses.
`causal_cross_entropy` is documented as *"weighted causal cross-entropy for
supervised updates through the RL path"* and uses the same `roll(-1)` convention as
`compute_logprobs_post`: `loss_mask[i]` weights the prediction `logits[i]` makes for
`input_ids[i + 1]`.

Net effect:

- **one** label convention for train and eval, on every `model_provider`
- no provider branching in the recipe — `use_next_token_labels` stops mattering for
  this path
- train and eval numbers become directly comparable, because they run the same loss
  code

To be precise about what this does and does not buy: it makes the convention
*uniform*, not automatic. Sequences still have to be rendered next-token aligned
(`next_token_labels=True`), because that is what drives the `loss_mask` that
`causal_cross_entropy` requires in `context` — `collate(with_rl_context=True)`
derives `loss_mask` from `sequence.labels`. The win is that the same tensors then
work for both endpoints, instead of one batch needing two variants.

The eval reduction is then:

```python
result = forward(client, job_id, kwargs)          # was forward_loss
logprobs = result["logprobs"]
n_supervised = (labels != IGNORE_INDEX).sum()
nll = -logprobs.sum() / n_supervised              # ignored positions come back as 0.0
```

**This is self-checking.** `causal_cross_entropy` is a token-weighted mean CE over
the same positions, so `nll` should reproduce the `avg_loss` that forward-backward
reports for the same batch, to floating-point tolerance. Worth asserting in the
recipe or a test — it catches a label-alignment mistake, a mis-assembled request
chunk, and a mis-decoded result in one comparison. A runnable version of that check
lives in `examples/forward_e2e.py` on
https://github.com/snowflake-eng/dss-client/pull/83 and can be ported.

## Two related things worth folding in

**1. The current `forward_loss` eval contaminates training.** `eval_nll` runs a full
backward per eval batch and never steps, immediately before `forward_backward_step`.
There is no `zero_grad()` anywhere in the dss-platform training worker — only
`engine.backward()` and `engine.step()`, and DeepSpeed only zeroes inside `step()`.
So the eval gradients accumulate and are folded into the next training update.
Switching the eval to a no-grad `forward` fixes that as a side effect, which also
means train loss will legitimately shift after the change — not a regression.

**2. On Cortex, `forward` returns logprobs, not a loss.** `forward_only` hardcodes
`processing = {"loss_fn": None, "post": ["compute_logprobs"]}` and ignores the
caller's `processing`, so a `{"loss_fn": "sft"}` (or `causal_cross_entropy`) body
yields no `avg_loss` / `metrics.loss` on that route. That makes
`arctic_platform/sft/examples/run_sft_ckpt_resume_demo.py::_eval_loss` — which reads
`metrics.get("loss", out.get("avg_loss"))` from `fwd_no_grad` — on-prem-only as
written. Anything reusing that helper against Cortex needs the logprob reduction
above instead. (I have not read the on-prem backend, so I am only claiming the
Cortex behavior here.)

## Dependencies

`forward` is not usable on Neutrino at all until these land:

- https://github.com/snowflake-eng/dss-platform/pull/410 — accepts the operation
  envelope, reassembles request chunks, honors `response_options` on `/forward`
- https://github.com/snowflake-eng/dss-client/pull/83 — chunks the payload and
  decodes the result

Arctic-Platform vendors its own copy of the client
(`cortex_training/cortex_training/client.py` on #88, whose `forward()` /
`_operation()` are byte-identical to the pre-fix dss-client), so that copy needs the
same port before any of this runs. Note also that
`arctic_platform/client/requests.py` forwards `batch` verbatim by design — *"the
caller must still match the target backend's data contract"* — so the label contract
stays a caller-side concern either way.

<sub>Correction to the first revision of this issue: it cited
`prepare_sft_request_labels` as the function doing the HF/Liger shift. That function
exists and has the same logic, but is only referenced from tests — the live path is
`shift_sft_labels_for_logits`, called from `dp_shard_batch`.</sub>

**Comment by sfc-gh-thonguyen** [comment]
Now measured on QA6, so the size of this trap is no longer hypothetical. Same live
training job (Qwen3-8B, bf16, ZeRO-2, `model_provider="huggingface"` — i.e. the
recipe's default), one batch, 84 supervised positions:

| what was sent to `/forward` | resulting NLL | vs `avg_loss` reference (5.132468) |
|---|---|---|
| next-token labels (`labels[t] == input_ids[t+1]`) | 5.132467 | **1e-6** — agrees |
| unshifted labels (what the recipe renders today) | 12.303701 | **off by 7.17 nats** |

So on an HF/Liger job the mistake more than doubles the reported loss. It is large
enough to notice if someone is watching the number, and small enough to look like a
plausible "eval is worse than train" if nobody is.

The agreement in the first row also confirms the self-check suggested above is worth
wiring in: it held to 1e-6 at both 27 and 84 supervised tokens, so a tolerance around
1e-3 catches a real misalignment with ~1000x margin.

Reproduction script (`examples/forward_e2e.py`) and the job body
(`examples/training-small-hf.json`) are on
https://github.com/snowflake-eng/dss-client/pull/83. Note the job needs
`flash_attention_3` on H200 — `flash_attention_2` failed to start.
